### ISSUE TITLE
Returning a tuple from a pure function causes Prusti to crash

### ISSUE THREAD

**Issue by vakaras** [viperproject/prusti-dev#298]
The following program:

```rust
use prusti_contracts::*;
#[pure]
fn dummy() { }
fn main() {}
```

gives a stacktrace:

```plain
hread 'rustc' panicked at 'not implemented: Tuple([])', prusti-viper/src/encoder/encoder.rs:1071:22
stack backtrace:                                                  
   0: rust_begin_unwind
             at /rustc/98d66340d6e63eda115afc8b0da1d87965881936/library/std/src/panicking.rs:495:5
   1: std::panicking::begin_panic_fmt
             at /rustc/98d66340d6e63eda115afc8b0da1d87965881936/library/std/src/panicking.rs:437:5
   2: prusti_viper::encoder::encoder::Encoder::encode_const_expr
             at /home/data/eth/rust/prusti-dev/prusti-viper/src/encoder/encoder.rs:1071:22
   3: prusti_viper::encoder::mir_encoder::MirEncoder::encode_operand_expr
             at /home/data/eth/rust/prusti-dev/prusti-viper/src/encoder/mir_encoder.rs:383:19
   4: <prusti_viper::encoder::pure_function_encoder::PureFunctionBackwardInterpreter as prusti_viper::encoder::mir_interpreter::BackwardMirInterpreter>::apply_statement
             at /home/data/eth/rust/prusti-dev/prusti-viper/src/encoder/pure_function_encoder.rs:931:48
   5: prusti_viper::encoder::mir_interpreter::run_backward_interpretation
             at /home/data/eth/rust/prusti-dev/prusti-viper/src/encoder/mir_interpreter.rs:74:13
   6: prusti_viper::encoder::pure_function_encoder::PureFunctionEncoder::encode_function
             at /home/data/eth/rust/prusti-dev/prusti-viper/src/encoder/pure_function_encoder.rs:81:25
   7: prusti_viper::encoder::encoder::Encoder::encode_pure_function_def
             at /home/data/eth/rust/prusti-dev/prusti-viper/src/encoder/encoder.rs:1223:37
   8: prusti_viper::encoder::encoder::Encoder::process_encoding_queue
             at /home/data/eth/rust/prusti-dev/prusti-viper/src/encoder/encoder.rs:1466:17
   9: prusti_viper::verifier::Verifier::verify
             at /home/data/eth/rust/prusti-dev/prusti-viper/src/verifier.rs:243:9
  10: prusti_driver::verifier::verify
             at /home/data/eth/rust/prusti-dev/prusti/src/verifier.rs:48:39
  11: <prusti_driver::callbacks::PrustiCompilerCalls as rustc_driver::Callbacks>::after_analysis::{{closure}}
             at /home/data/eth/rust/prusti-dev/prusti/src/callbacks.rs:82:17
  12: rustc_interface::passes::QueryContext::enter::{{closure}}
             at /home/vakaras/.rustup/toolchains/nightly-2020-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/compiler/rustc_interface/src/passes.rs:729:42
  13: rustc_middle::ty::context::tls::enter_context::{{closure}}
             at /home/vakaras/.rustup/toolchains/nightly-2020-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/compiler/rustc_middle/src/ty/context.rs:1719:50
  14: rustc_middle::ty::context::tls::set_tlv
             at /home/vakaras/.rustup/toolchains/nightly-2020-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/compiler/rustc_middle/src/ty/context.rs:1703:9
  15: rustc_middle::ty::context::tls::enter_context
             at /home/vakaras/.rustup/toolchains/nightly-2020-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/compiler/rustc_middle/src/ty/context.rs:1719:9
  16: rustc_interface::passes::QueryContext::enter
             at /home/vakaras/.rustup/toolchains/nightly-2020-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/compiler/rustc_interface/src/passes.rs:729:9
  17: <prusti_driver::callbacks::PrustiCompilerCalls as rustc_driver::Callbacks>::after_analysis
             at /home/data/eth/rust/prusti-dev/prusti/src/callbacks.rs:48:9
  18: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  19: rustc_span::with_source_map
  20: rustc_interface::interface::create_compiler_and_run
  21: rustc_span::with_session_globals
```

Since `()` is `Copy`, this code should be legal.

cc @cmatheja

**Comment by Pointerbender** [comment]
I have another example which seems to be related:

```rust
#[derive(Clone, Copy)]
pub struct A;

impl A {
    #[pure]
    pub fn get(&self) -> usize {
        0
    }
}

#[pure]
pub fn test(a: A) {
    a.get();
}
```

Error:

```
thread 'rustc' panicked at 'internal error: entered unreachable code: Tuple([])', prusti-viper/src/encoder/type_encoder.rs:206:22
stack backtrace:
   0: rust_begin_unwind
             at /rustc/5ecc8ad8462574354a55162a0c16b10eb95e3e70/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt
             at /rustc/5ecc8ad8462574354a55162a0c16b10eb95e3e70/library/core/src/panicking.rs:103:14
   2: prusti_viper::encoder::type_encoder::TypeEncoder::get_integer_bounds
   3: prusti_viper::encoder::type_encoder::TypeEncoder::encode_bounds
   4: prusti_viper::encoder::pure_functions::encoder::PureFunctionEncoder::encode_function_given_body
   5: prusti_viper::encoder::pure_functions::encoder::PureFunctionEncoder::encode_function
   6: <prusti_viper::encoder::encoder::Encoder as prusti_viper::encoder::pure_functions::interface::PureFunctionEncoderInterface>::encode_pure_function_def
   7: prusti_viper::encoder::encoder::Encoder::process_encoding_queue
   8: prusti_viper::verifier::Verifier::verify
   9: rustc_interface::passes::QueryContext::enter
  10: <prusti_driver::callbacks::PrustiCompilerCalls as rustc_driver::Callbacks>::after_analysis
  11: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  12: rustc_span::with_source_map
  13: scoped_tls::ScopedKey<T>::set
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by fpoli** [comment]
This might be fixed by #732.
